refactor(sdk-metrics-base): meter shared states

legendecas opened the pull request:
## Which problem is this PR solving?

Refactors meter states so that we can safely add more meter internal fields like async instrument callback registrations.

Also hides the public `Meter.collect` method in the shared state interface so that it can only be exposed through `MetricReader`.

This is extracted from https://github.com/open-telemetry/opentelemetry-js/pull/2785 as part of the async callback re-work.

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] MeterSharedState.collect

## Checklist:

- [x] Followed the style guidelines of this project
- [x] Unit tests have been added

legendecas commented on tsconfig.es5.json:
```diff
@@ -1,6 +1,7 @@
   "compilerOptions": {
-    "target": "es5"
+    "target": "es5",
+    "downlevelIteration": true
```
typescript is complaining on https://github.com/open-telemetry/opentelemetry-js/blob/main/experimental/packages/opentelemetry-sdk-metrics-base/src/state/MetricStorageRegistry.ts#L31 that ES6 iterators need to be transpiled to work on ES5.

legendecas replied:
Submitted #2823 to fix this.

legendecas commented:
This was closed accidentally. Reopened, and rebased.

pichlermarc commented on experimental/packages/opentelemetry-sdk-metrics-base/test/export/TestMetricExporter.ts:
```diff
@@ -0,0 +1,39 @@
+import { ExportResult, ExportResultCode } from '@opentelemetry/core';
+import { AggregationTemporality, PushMetricExporter, ResourceMetrics } from '../../src';
+
+export class TestMetricExporter implements PushMetricExporter {
+  resourceMetricsList: ResourceMetrics[] = [];
+  export(resourceMetrics: ResourceMetrics, resultCallback: (result: ExportResult) => void): void {
+    this.resourceMetricsList.push(resourceMetrics);
+    process.nextTick(() => resultCallback({ code: ExportResultCode.SUCCESS }));
+  }
+
+  async forceFlush(): Promise<void> {}
+  async shutdown(): Promise<void> {}
+
+  getPreferredAggregationTemporality(): AggregationTemporality {
+    return AggregationTemporality.CUMULATIVE;
+  }
+}
```
Non-blocking for me, but there's another (a bit more elaborate) `TestMetricExporter` in `PeriodicExportingMetricReader.test.ts` that could be combined with this one.

It's probably a bit out of scope for this PR, but I can open a follow up PR to cut down on some duplication. :slightly_smiling_face:

legendecas replied:
Yeah, that one contains lots of custom functionalities that we don't have a common use case outside of `PeriodicExportingMetricReader.test.ts` so I didn't merge that one.

pichlermarc replied:
Alright, good to know, thanks. Then we can leave this as is. :slightly_smiling_face:

pichlermarc reviewed (approved):
Looks great! :slightly_smiling_face:

dyladan commented on experimental/packages/opentelemetry-sdk-metrics-base/test/export/TestMetricExporter.ts:
```diff
@@ -0,0 +1,39 @@
+  async forceFlush(): Promise<void> {}
+  async shutdown(): Promise<void> {}
+
+  getPreferredAggregationTemporality(): AggregationTemporality {
```
I think the concept of "preferred temporality" is removed from the spec. Temporality is now supposed to be decided by the MetricReader which can take a callback which decides temporality. For now this is fine and not blocking this PR, but I wanted to make sure we don't forget about it.

legendecas replied:
I created #2864 to track this.

dyladan reviewed (approved):
This looks mostly like a "pure" refactoring to me and no new functionality right? LGTM

dyladan commented:
@open-telemetry/javascript-approvers please take a look at this if you have time

seemk reviewed (approved):
My moral support :slightly_smiling_face:

legendecas commented:
@open-telemetry/javascript-approvers this pr still needs one more review, I'd appreciate it if anyone can take a look at this :) I've updated it on top of the main branch and added a changelog entry.

dyladan commented:
I'll bug everyone in the SIG meeting :)

legendecas commented:
@open-telemetry/javascript-approvers merged the latest main branch and resolved conflicts. we still need one more review on this PR :)

legendecas commented:
Unexpected push... reverted to the previous version https://github.com/open-telemetry/opentelemetry-js/commit/03fa81f0c99f68cfef70c6a5d75312d6c32231a7.

We are still waiting for another review to merge :-)

legendecas commented:
@vmarchaud thank you
